pkg/controller/complianceremediation/complianceremediation_controller.go

```
@@ -134,32 +134,105 @@ func (r *ReconcileComplianceRemediation) reconcileRemediation(instance *compv1al
-			return common.NewNonRetriableCtrlError("Unable to create ComplianceRemediation. Make sure the CRD is installed: %s", err)
+
+	if errors.IsForbidden(err) {
+		return common.NewNonRetriableCtrlError(
+			"Unable to get fix object from ComplianceRemediation. "+
+				"Please update the compliance-operator's permissions: %s", err)
+	} else if meta.IsNoMatchError(err) {
+		// If the kind is not available in the cluster, we can't retry
+		return common.NewNonRetriableCtrlError(
+			"Unable to get fix object for ComplianceRemediation. "+
+				"Make sure the CRD is installed: %s", err)
+	} else if errors.IsNotFound(err) {
+		if instance.Spec.Apply {
+			return r.doCreateGenericRemediation(obj, found, objectLogger)
 		}
-		// TODO(jaosorior): If the object is found, should we update it?
+
+		objectLogger.Info("The object wasn't found, so no action is needed to unapply it")
+		return nil
+	} else if err != nil {
 		return err
-	} else {
-		logger.Info("Remediation will be unapplied", "ComplianceRemediation.Name", instance.Name)
 	}
 
-	// unapply remediation
-	if err == nil {
-		return r.client.Delete(context.TODO(), obj)
-	} else if errors.IsNotFound(err) {
+	if instance.Spec.Apply {
+		return r.doPatchGenericRemediation(obj, found, objectLogger)
+	}
+
+	return r.doDeleteGenericRemediation(obj, found, objectLogger)
+}
+
+func (r *ReconcileComplianceRemediation) doCreateGenericRemediation(remObj *unstructured.Unstructured, foundObj *unstructured.Unstructured, logger logr.Logger) error {
+	logger.Info("Remediation will be created")
+	compv1alpha1.AddRemediationAnnotation(remObj)
+
+	createErr := r.client.Create(context.TODO(), remObj)
+
+	if errors.IsForbidden(createErr) {
+		// If the kind is not available in the cluster, we can't retry
+		return common.NewNonRetriableCtrlError(
+			"Unable to create fix object from ComplianceRemediation. "+
+				" Please update the compliance-operator's permissions: %s", createErr)
+	} else if errors.IsAlreadyExists(createErr) {
```

If we were asked to create an object but it exists, isn't it better to return a retriable error and let the loop cycle back into the patch branch?

---

that's a good point.